druid-shell/src/mac/menu.rs

```diff
@@ -13,17 +13,105 @@
+use cocoa::appkit::{NSMenu,NSMenuItem, NSApplication};
 
-pub struct Menu;
+struct MenuItemProxyClass(*const Class);
+unsafe impl Sync for MenuItemProxyClass {}
 
+lazy_static! {
+    static ref MENU_ITEM_PROXY_CLASS: MenuItemProxyClass = {
+        unsafe {
+            let mut decl = 
+                ClassDecl::new("DruidMenuItemProxy", class!(NSObject))
+                .expect("Menu class defined");
+            decl.add_ivar::<u32>("menu_id");
+            decl.add_method(
+                sel!(trigger),
+                trigger as extern "C" fn (&Object, Sel)
+            );
+            extern "C" fn trigger(this: &Object, _: Sel) {
+                unsafe {
+                    let menu_id: u32 = *this.get_ivar("menu_id");
+                    println!("triggered menu item with id {}", menu_id);
+                }
+            }
+            MenuItemProxyClass(decl.register())
+        }
+    };
+}
+pub struct Menu {
+    pub menu: id,
+}
+fn make_basic_menu_item(id: u32, text: &str, key: &str) -> id {
+    unsafe {
+        NSMenuItem
+            ::alloc(nil)
+            .initWithTitle_action_keyEquivalent_(
+                NSString::alloc(nil).init_str(text),
+                sel!(trigger),
+                NSString::alloc(nil).init_str(key)
+            ).autorelease()
+    }
+}
+
+fn make_proxy(id: u32) -> id {
+    unsafe {
+        let mut target: id = msg_send![MENU_ITEM_PROXY_CLASS.0, alloc];
+        (*target).set_ivar("menu_id", id);
+        target.autorelease()
+    }
+}
+fn make_menu_item(id: u32, text: &str, key: &str) -> id {
+    unsafe {
+        let target = make_proxy(id);
+        let mut menu_item = make_basic_menu_item(id, text, key);
+        msg_send![menu_item, setTarget: target];
+        menu_item
+    }
+}
 impl Menu {
     pub fn new() -> Menu {
-        Menu
+        unsafe {
+            Menu { menu: NSMenu::alloc(nil) }
+        }
     }
 
-    pub fn add_dropdown(&mut self, menu: Menu, text: &str) {}
+    pub fn add_dropdown(&mut self, menu: Menu, text: &str) {
+        unsafe {
+            let menu_item = NSMenuItem::new(nil).autorelease();
+            menu_item.setSubmenu_(menu.menu);
+            self.menu.addItem_(menu_item);
+        }
+    }
 
-    pub fn add_item(&mut self, id: u32, text: &str) {}
+    pub fn add_item(&mut self, id: u32, text: &str, key: &str) {
+        unsafe {
+            let menu_item = make_menu_item(id, text, key);
+            self.menu.addItem_(menu_item);
+        }
+    }
 
-    pub fn add_separator(&mut self) {}
+    pub fn add_separator(&mut self) {
+        unsafe {
+            let sep = id::separatorItem(self.menu);
+            self.menu.addItem_(sep);
+        }
+    }
+}
+
+impl Default for Menu {
+    fn default() -> Menu {
+        let mut menu = Menu::new();
+        let mut submenu = Menu::new();
+        submenu.add_item(1, "Quit", "q");
+        submenu.add_separator();
+        submenu.add_item(1, "Also Quit", "q");
```

Comment:
I understand the desire to test but probably just the one quit is enough :)

Reply:
It's only there to show the separator works (the separator will not render if you remove the second item)

Reply:
I'm not actually sure what a sensible default is here. My initial thought was None, but actually we probably are better to just defaulting to Quit. That said, I have no real idea how it all threads through the rest of the library yet.

Reply:
Yeah, defaulting to quit is pretty reasonable, I think. It's different than Windows, where it's common to have a window that doesn't have any menus at all.